### DISCUSSION TITLE
socmaxtemp question...

### DISCUSSION THREAD

**Discussion by pottendo** [rsta2/circle#428] [General]
Hi,
on a RPi4 I tried to monitor/control the temperature when throttling should happen.
To test this cmdline.txt;
```
console=serial0,115200 socmaxtemp=60
```

The code:
```
    unsigned tmask = SystemStateUnderVoltageOccurred | SystemStateFrequencyCappingOccurred |
					 SystemStateThrottlingOccurred | SystemStateSoftTempLimitOccurred;
	CPUThrottle.RegisterSystemThrottledHandler(tmask, monitorhandler, nullptr);
	// CPUThrottle.DumpStatus(true);
	log("Maximum temp set to %d, dynamic adaption%spossible, curret freq = %dMHz (max=%dMHz)", 
			CPUThrottle.GetMaxTemperature(), 
			CPUThrottle.IsDynamic() ? " " : " not ",
			CPUThrottle.GetClockRate() / 1000000L, 
			CPUThrottle.GetMaxClockRate() / 1000000L);

	CMachineInfo *mi = CMachineInfo::Get();
	if (mi)
	{
		TMachineModel model = mi->GetMachineModel();
		if (((model == MachineModel4B) ||
			 (model == MachineModel400) ||
			 (model == MachineModelCM4)) &&
			(CPUThrottle.SetSpeed(CPUSpeedMaximum, true) != CPUSpeedUnknown))
		{
			log("maxed freq to %dMHz", CPUThrottle.GetClockRate() / 1000000L);
		}
	}

```
which results in:
```
00:00:02.23 pottendo-log:: Maximum temp set to 85, dynamic adaption possible, curret freq = 600MHz (max=1500MHz)
00:00:02.24 pottendo-log:: maxed freq to 1500MHz

```

So it seems that `socmaxtemp=...` in `cmdline.txt` has no effect - `CPUThrottle.GetMaxTemperature()` still delivers 85C, so my handler is not called as I've not reached 85 degrees.

What am I doing wrong?
thx, pottendo

**Comment by rsta2** [comment]
You have to call CPUThrottle.Update() in your application's main loop from 
time to time (at least all 4 seconds).

**Reply by pottendo** [reply]
Ah, OK. 
Im calling `SetOnTemperature(void)` already. Didn't read the docs further down... silly me! 
Is the Core significant? My Monitor runs on Core3 - which probably could be changed.
Thx pottendo

**Comment by rsta2** [comment]
SetOnTemperature() does not call the registered callback. This is only done by 
Update(). I assume that it can be called from Core 3, but the callback will be 
called on this core too then.

**Reply by pottendo** [reply]
Hi,
somehow it works:
here the monitor code, which is launched on core 3:
```
void CKernel::run_tempmonitor(void)
{
    unsigned tmask = SystemStateUnderVoltageOccurred | SystemStateFrequencyCappingOccurred |
					 SystemStateThrottlingOccurred | SystemStateSoftTempLimitOccurred;
	CPUThrottle.RegisterSystemThrottledHandler(tmask, monitorhandler, nullptr);
	CPUThrottle.Update();

	while (true) {
		if (CPUThrottle.SetOnTemperature() == false)
			log("temperature monitor failed...");
		MsDelay(5 * 1000);
		if (!CPUThrottle.Update())
			log("CPUThrottle Update failed");
		log("Temp %dC (max=%dC), dynamic adaption%spossible, current freq = %dMHz (max=%dMHz)", 
			CPUThrottle.GetTemperature(),
			CPUThrottle.GetMaxTemperature(), 
			CPUThrottle.IsDynamic() ? " " : " not ",
			CPUThrottle.GetClockRate() / 1000000L, 
			CPUThrottle.GetMaxClockRate() / 1000000L);
	}
}

```
and the output:
```
00:00:37.49 pottendo-log:: Temp 55C (max=85C), dynamic adaption possible, current freq = 1500MHz (max=1500MHz)
00:00:42.50 pottendo-log:: Temp 54C (max=85C), dynamic adaption possible, current freq = 1500MHz (max=1500MHz)
00:00:47.51 pottendo-log:: Temp 55C (max=85C), dynamic adaption possible, current freq = 600MHz (max=1500MHz)
00:00:52.51 pottendo-log:: Temp 52C (max=85C), dynamic adaption possible, current freq = 600MHz (max=1500MHz)
00:00:57.52 pottendo-log:: Temp 54C (max=85C), dynamic adaption possible, current freq = 600MHz (max=1500MHz)
00:01:02.53 pottendo-log:: Temp 52C (max=85C), dynamic adaption possible, current freq = 1500MHz (max=1500MHz)
00:01:07.54 pottendo-log:: Temp 53C (max=85C), dynamic adaption possible, current freq = 1500MHz (max=1500MHz)

```

As one can see the throttling happens. I've forced it with `maxsoctemp=55` in `cmdline.txt`.

However, the registered callback is _not_ called in my case.
Also I had expected `CPUThrottle.GetMaxTemperature()` would give you the supervised upper threshold - in my case `55` from `cmdline.txt`. But this may have been a wrong assumption.

Anyway, problem solved in principle as the throttling shows the desired effect (temp goes down) -> consequently my app fails then due to too slow responses, but that's a different story... ;-)

thx, pottendo

PS: 
Here's the code of the monitor - spiced with some logs to see if it's called.
```
void monitorhandler(TSystemThrottledState CurrentState, void *pParam)
{
	Kernel.log("%s - state = 0x%04x", __FUNCTION__, CurrentState);
	if (CurrentState & SystemStateUnderVoltageOccurred) {
		Kernel.log("%s: undervoltage occured...", __FUNCTION__);
	}
	if (CurrentState & SystemStateFrequencyCappingOccurred) {
		Kernel.log("%s: frequency capping occured...", __FUNCTION__);
	}
	if (CurrentState & SystemStateThrottlingOccurred) {
		Kernel.log("%s: throttling occured to %dMHz", __FUNCTION__, CPUThrottle.GetClockRate() / 1000000L);
	}
	if (CurrentState & SystemStateSoftTempLimitOccurred) {
		Kernel.log("%s: softtemplimit occured...", __FUNCTION__);
	}
}
```

**Comment by rsta2** [comment]
The CPU throttle callback is triggered from the firmware, when one of the 
conditions occurs, which can be selected with the mask parameter. The 
SystemStateSoftTempLimitOccurred condition uses the the maximum SoC 
temperature here, which is reported by the firmware (85°C). This is different 
from the value, which is set with socmaxtemp=, which is enforced by Circle 
CCPUThrottle itself, not by the firmware.

So when the callback is not called, no one of the system throttled conditions 
has been triggered. I have to admit, that I haven't tested this for a while. 
Changes to the firmware API sometimes occur.

You don't need to call CPUThrottle.SetOnTemperature(), if you call 
CPUThrottle.Update(). Both do nearly the same, but Update() also checks the 
system throttled state and ensures, that it is not called too frequently. If 
Update() is called, when the last call to it was later than 4 seconds ago, the 
call is simply ignored. So you can call Update() all the time, without the 
need to care about the timing.

**Reply by pottendo** [reply]
Thx for the clarification, pottendo
